### ISSUE TITLE
Improve handling of domain instances across multiple deployment plans

### ISSUE THREAD

**Issue by j26151** [ZeligsoftDev/CX4CBDDS#400]
# **Issue and tracking information**
We are looking to make a sizable change to domain modeling in CX.  We have a current issue that we handle manually and that is if we have say 1 domain with 5+ different sets of node->IP mappings and they exist in 10 different deployment plans, we have to ensure that all of the domains mappings stay in sync between all of the deployment plans. Additionally, any one of those deployment plans, when generated, overwrite the CDDs from the others. Here is what we think would solve this problem:

There would be a new first-class model entity to define & generate one or more CDD files.  Maybe called a Domain Definition rather than a Deployment Plan, with a new icon.  The new Domain Definition entity would not require the right side deployment plan mapping pane in the main window that shows component->process->computer mappings.  Just left side pane with the ability to configure one or more CDDs and set per-computer & domain properties for each.  The new Domain Definition model element would generate CDD files, and that's it.  No CDPs.

The current Deployment Plan model entity would retain its current deployment plan "editor" 2-pane main window to set properties on component container processes & assembly trees, and support drag-drop of them onto the right pane to defined the component->process->computer mappings & generate a single CDP.  The current Deployment Plan entity would require a Domain Definition entity as a dependency in its left pane, to support doing the process -> computer mappings for a deployment.  But the Deployment Plan editor would no longer let users set properties on Domains (the separate Domain Definition model entities are for that), and it would only generate a CDP file - not CDD file(s).

### Developer's Actual time spent on fix (hours)  

# **Issue reporter to provide a detailed description of the issue in the space below**

**Comment by eposse** [comment]
Hi. I have a few questions about this issue.

First, I want to be sure I understand the problem, how you are currently dealing with it, and your proposed solutions.

You say to consider the situation where "we have say 1 domain with 5+ different sets of node->IP mappings and they exist in 10 different deployment plans". How exactly are you modelling this? I'm assuming that you are dealing with this as described in #396, where you have *one* Domain element (e.g. ExamplesDomain), you drag-and-drop them into the Configure pane (i.e. the left pane) of each of the 10 Deployment Plans, 5+ times, and each such "instance" in that pane (ExamplesDomain_1, ExamplesDomain_2, etc.) is edited with its own node->address mapping, as shown in the BPS example in #396. Is this correct?

Then you add "Additionally, any one of those deployment plans, when generated, overwrite the CDDs from the others." If my understanding is correct, You have dragged-and-dropped "ExamplesDomain" several times into, say, DeploymentPlan_1, DeploymentPlan_2, etc., and then, for example, changed the address of node1 on ExamplesDomain_2 in DeploymentPlan_1, and also changed the address for the same node1 on ExamplesDomain_2 as well in DeploymentPlan_2, and when you generate the CDDs for DeploymentPlan_2, they overwrite those of DeploymentPlan_1. Is this correct?

If my understanding above of the current situation is correct, your proposed solution of a new "Domain Definition" element with its own editor would mean that this new element would have as children, one or more Domain elements with node -> address mappings? For example, we would have something like

```
ExamplesDomainDefinition
|
+- ExamplesDomain_1
|   |
|   +- node1: (maps to addr1)
|   +- node2: (maps to addr2)
|
+- ExamplesDomain_2
|   |
|   +- node1: (maps to addr3)
|   +- node2: (maps to addr4)
|
...
```

Which would be editable in a new (single pane) editor, and then your DeploymentPlan_1 and DeploymentPlan_2 would have a (dependency) reference to ExamplesDomainDefinition on its Configure (left) pane, where it would *not* be editable in the DeploymentPlan editor. Is this correct?

Furthermore, we would separate the descriptor generation so that CDDs are only generated from a Domain Definition element, while CDPs (and only CDPs, no CDDs) are generated from Deployment Plan elements, is this correct?

**Comment by MHayman5** [comment]
It sounds like you understand the issue & proposed solution.  "Yes" to all the "is this correct" questions.

**Comment by ysroh** [comment]
Do we allow one to many mapping between a domain definition to domains? Or one-to-one mapping? If you have multiple domains in a domain definition, you will see multiple domains in the deployment configure pane, and is that your expected workflow?

**Comment by MHayman5** [comment]
One to many.  A single common domain definition should just provide node names and basic node type attributes that can be set differently when instantiated multiple times in the configure pane for a domain editor.

All the deployment plan editor needs relative to a domain is the common domain definition type that defines the node names, so it can support mapping container processes that it adds to node names.

**Comment by eposse** [comment]
Another question: the new "Domain Definition" element seems to be a "container" for one or more concrete "Domains" which will provide specific node->address mappings. Are there any constraints on the Domains contained within a Domain Definition? For example, should all Domain elements included in a Domain Definition have the same nodes?

**Comment by MHayman5** [comment]
By "same nodes" do you mean the same computer node types with common settable attributes/attributes, just instantiated perhaps in different numbers (with alt graphical interconnects)?  If so, we could live with that since all we really use in a Domain Definition, or Domain, or whatever, is the computer node logical name, and the ability to set an  IP address/port attribute/property for each node.  The current common computer node type easily handles that already.

However, ideally an OMG deployment & config (D&C) spec Domain could include a variety of general purpose processor (GPP) computer or alt processing node types (GPU, FPGA, custom device...) instantiated within it, with different attributes or properties settable on each.  For instance, a distributed computing GPP computer target domain populated by let's say 4 Intel x86 CPUs, and 8 ARM embedded System-On-Chip (SoC) CPUs, all interconnected on an Ethernet network.  The x86 vs. ARM compute nodes might have different properties in a Domain like this.

Again though, in our use of the CX tool to model Domains they're all just network interconnected "computers" that run software.  So a common general purpose compute node type for all with settable name and IP address/port properties is adequate.

**Comment by elder4p** [comment]
@MHayman5  & @j26151:

Because this is such a big change, I’m stepping in to make sure we have a common understanding of the problem and the proposed solution. Bear with me if I’m repeating details already discussed.

**The fundamental problem**: This is no single model element representing a CDD file. 

**Current situation**: Domain elements are defined once. Each Deployment Plan references one or more Domain objects (as a’ Domain Instance’, not an explicit meta model type). CDD deployment attributes attribute (e.g. node addresses) are edited in the Deployment Plan editor by modifying values on each ‘Domain Instance’. When the Deployment Plan is generated, a CDP file is produces as well as one CDD file for each Domain Instance created in the plan.

**The pain point**: Multiple deployment plans can used the same Domain and (logically) define the same ‘Domain Instance’. However, because there is no single model element representing this shared domain instance, the CDD file gets generated from each Deployment Plan. This repetition creates a risk that deployment plans will inconsistently describe what will eventually become a CDD file, leaving the resulting CDD dependent on code generation order.


Assuming the above accurately reflects the current situation, let me turn to the solution. We need to express the new meta-model concepts, and then to walk you through the new user work flow.

**My fundamental goal**: have a single model element corresponding to each CDD file. It would be an explicit meta model type.

**New Meta-Model type**: ‘**Domain Deployment**’ representing a single CDD file to generate.

This is a little different from your suggestion of a ‘Domain Definition’ object, but not inconsistent. I view your ‘Domain Definition’ proposal as representing a collection related CDD files; that is, as containing ‘Domain Deployment’ objects.

**New Meta-Model type**: ‘**Domain Definition**’ containing a collection of related ‘Domain Deployment’ elements.

**New meta-model Attribute:** ‘Domain Definition’ has a multivalued containment attribute ‘**deployments**’ of type ‘Domain Deployment’.

Furthermore, I imagine that ‘Domain Definition’ will serve two roles:
* Express domain structure to a Deployment Plan. From what I can see, a CDP file (and thus a Deployment Plan) only needs to know the display names of NodeInstances contained in the Domain Definition.  In the ‘new world’, a Deployment Plan will reference a single ‘Domain Definition’, from which all node display names can be obtained.
* Be an anchor point for defining and generating CDD files and their deployment properties (principally their node addresses, I think).

From the first point, I deduce that all CDD files within a Domain Definition must share a common set of node names. To avoid repetition, these should be defined once, in the Domain Definition itself.

**New meta-model Attribute**: ‘Domain Definition’ has a multivalued containment attribute ‘**node**’ of type ‘NodeInstance’. (This is intentionally identical to how the current ‘Domain’ type is defined.) 
From the perspective of the Deployment Plan editor and CDP generation, the above is all that is required of a Domain Definition.

Exactly how we model ‘Domain Definition’ and ‘Domain Deployment’ depends on how much flexibility you require in defining related CDD files.

Your current CX usage pattern implies a very restrictive model:
* The structure of Domain Definition would be identical to structure of the current Domain concept. Namely, it contains NodeInstances, but also InterconnectInstances, BridgeInstances, AssemblyConnectors, etc.
* Meanwhile, a Domain Deployment would be restricted to specifying the CDD file name and whatever properties values can edited in the Config pane of the Deployment Editor.

In particular, adopting this restrictive model would mean two CDD files generated from the same Domain Definition would have identical structure and element names (NodeInstance, BridgeInstances, connectors, etc…); the only differences would be in deployment property values.

The extremely loose coupling between CDP and CDD files means there would be work arounds available should this be too restrictive, but CX would not help you in maintaining consistency (without further work on our part). You could, for example, create more than one Domain Definition that, so long as they maintained consistent node names, would create CDD files usable by a single CDP. However, the Deployment Plan would only refer to one of these Domain Definitions. Maintaining consistency with any others would be a manual exercise.

The above would be the easiest meta model change for us to make. **Does it supply all the flexibility you need?**

Assuming all of the above is OK with you, here’s how I see CX and your work flows changing.

1. Developers would have a new Add CX option, **Add CX -> Domain Definition**. Functionally, it would behave identically to Add CX -> Domain. Users would continue to add Nodes and other structure via a Component Diagram. The only difference from Domain would be that Domain Definition objects would have addition Add CX and code generation options, and the Deployment Editor would treat Domain Definition objects slightly differently.
2. The Deployment Editor would allow a Domain Definition to be dragged into the Config pane of the editor, just as a Domain can be.
3. The Deployment Editor would allow a Domain Definition to be dragged to the Deployment (right hand) pane, just as a Domain can be.
4. The Deployment Editor would allow component and processes to be assigned to Domain Definition nodes just as with Domain nodes.
5. The Deployment Editor would not allow property editing of any configured Domain Definition or any of its children. (It may not even display these properties.)
6. The Deployment Editor would not allow the same Domain Definition to be added to the deployment plan Config pane more than once.
7. **Q**: Should the Deployment Editor remove any existing Domain objects from the Config pane when a Domain Definition is added?
8. **Q**: Is there any reason for a Deployment Plan to be configured with more than one Domain Definition? If not, the Deployment Editor should either prevent a second Domain Definition (or Domain) from being added, or, replace the existing object with the newly added one. (Preventing a ‘second add’ is easier.)
9. The code generation action for the Deployment Plan would continue to generated CDD files from Domain objects in the plan, but would not generate CDD files for configured Domain Definition(s) in the plan.

Although the ticket description envisages a new editor for Domain Definition objects, I believe this can be done using only the Add CX menu and the standard properties view as follows:

9. The Add CX menu for Domain Definition would include a new choice (compared to Domain) **Add CX -> Domain Deployment (CDD)**. This would create a child element under the Domain Definition. Ideally, it would not appear any component diagram for the Domain Definition (but I have to verify we can prevent this). The icon for this new element would be some variation of a standard file icon (or a cdd file icon, if such an icon exists).
10. A user selecting a Domain Deployment in the Model explorer would see, in the CX Properties View, a tree of Nodes (and Bridges, etc), and their configuration properties. The user would be able to edit and save those property values via this view.
11. The CX Properties view for a Domain Deployment would also contain the display name of the Domain Deployment. This would get the base name of the generated CDD file (i.e. the name without the .cdd extension). Q: is this sufficient? Another possibility is that the Domain Definition provide part of the base name and the Domain Deployment provide a second part. E.g. A Domain Definition named ‘ArtGallery’ might have two Domain Deployments named ‘intel’ and ‘arm’ that would result in cdd files ‘ArtGallery_intel.cdd’ and ‘ArtGallery_arm.cdd’.
12. A user right clicking over a Domain Definition in the Model Explorer view would see a CX generation menu item allowing them to generate all CDD files (for the definition).
13. A user right clicking over a Domain Deployment in the Model Explorer view would see a CX generation menu item allowing them to generated the specific CDD file.

Note that none of the above precludes the continued use the existing Domain elements. I believe this continued support is essential for a smooth transition to the new capability. 

We might, in the future, consider adding actions that would transform a Domain into a Domain Definition, although depending on how widely distributed the references to Domains are, this may not be practical. We could, at least, add a model validation warning (on Domain elements) indicating that there is a new and better way model domains.

Apologies for being so long winded, but I feel it is important we fully understand this before we get started on implementing. Comments?

**Comment by eposse** [comment]
@MHayman5 @j26151 @J17359 

I've opened PR #417 implementing this enhancement. You can download the build from [here](https://github.com/ZeligsoftDev/CX4CBDDS/actions/runs/2789421975).

This implements the concepts described by Paul above:

* A new "DomainDefinition" concept
* A new "DomainDeployment" concept

A DomainDefinition, like the old "Domain" can have node/bridge/interconnect instances, whose properties' default values can be edited in the CX tab of the Properties view. It can also have one or more DomainDeployments, which automatically inherit the node/bridge/interconnect instances from the DomainDefinition, and whose properties values can also be edited in the CX tab of the Properties view.

Each DomainDeployment corresponds to a CDD file. A DomainDefinition defines the common deployment parts (node/bridge/interconnect instances) for a set of DomainDeployments, with their respective properties (such as node addresses) and their default values.

To create a new DomainDefinition, right-click on a package, and select "Add CX -> Domain Definition". It has an icon like "Domain" but green. It will automatically add a diagram.

To add node/bridge/interconnect instances, you can either right-click the DomainDefinition and select "Add CX -> Node Instance", etc. or you can drag and drop them from the palette into the diagram.

To add a new DomainDeployment to a DomainDefinition, right-click the DomainDefinition and select "Add CX -> Domain Deployment". Its icon is like the Domain icon but pinkish.

When a new DomainDeployment is added, all the node/bridge/interconnect instances in the DomainDefinition are automatically inherited by the DomainDeployment. Furthermore, if new node/bridge/interconnect instances are added (or deleted) to (resp. from) the DomainDefinition, they are automatically added (resp. deleted) from all the DomainDeployments inside the DomainDefinition. 

To edit the properties of a DomainDeployment's node/bridge/interconnect instance, click on the instance inside the deployment in the Model Explorer, and select the CX tab in the Properties view. Then you can edit the properties in the same way as they were editable from the DeploymentPlan.

You can drag and drop a DomainDefinition (but not a DomainDeployment) into the "Configure" (left) pane of a Deployment Plan, and from there you can drag it to the "Deploy" (right) pane. You can allocate processes to nodes in the same way as with old Domains. Note that if you select a Node under a DomainDefinition in the DeploymentPlan editor, its properties won't be editable. You can only edit the properties in the DomainDeployments or DomainDefinitions themselves.

Code generation is separate. You can generate individual CDDs by right-clicking on a DomainDeployment and selecting "Generate -> CDD file". You can generate all CDDs of all DomainDeployments in a DomainDefinition by right-clicking the DomainDefinition and selecting "Generate -> All CDD files". Similarly, you can generate all CDDs in the model by selecting the same menu from the root of the model.

For CDPs its essentially the same: you generate only CDPs by right-clicking on a DeploymentPlan in the Model Explorer and selecting "Generate -> CDP file", or right-clicking the root element of the model and selecting "Generate -> All CDP files".

We kept the old generation menus "Generate Descriptors" and "Generate All Descriptors", while your users migrate to the new approach. These old menus work as before, generating both a CDP and CDDs for each DeploymentPlan, although the CDDs will be generated only for the old Domains and not for any new DomainDefinition. 

Let us know if/when you want the old menus removed or if you have any other questions.

**Comment by j26151** [comment]
Sorry for taking so long to get around to reviewing this. This looks really good, thank you for your hard work here. We have a few small fixes and changes that we would like as well as a (hopefully) small feature improvement.

To start with some of the small changes:

1. Copy and paste of a DomainDeployment doesn't appear to work completely. It creates a new DomainDeployment with 0 entries, so we can't modify the nodes properties or anything.
1. We would actually the following menu settings for generating new artifacts.
2.1. The top level options should remain "Generate Descriptor Files" and there should be no options for "CDP" or "CDD" files. This would generate all CDD and CDP files from all types of Domains and deployment plans
2.2. If an individual Deployment plan is right clicked on, it should also say "Generate Descriptor Files". This would generate all CDD files associated with the deployment plan as well as the CDP file
2.3. If and individual DomainDefinition was right clicked, it should say "Generate CDD Files".  This would generate just the CDD files for this definition
1. When creating a new CX element the old "Domain" option should be changed to "Domain (Deprecated)". This feature will remain for a while. Our programs are slow at adoption so we want to make sure its clear that its deprecated, but continue allowing the feature.

As for the feature improvement. I received a really good use case that we would like to address now if possible. This use case is where a model, most likely a dynamic pathmapped model would have a DomainDefinition defined and some common DomainDeployments defined under it. Then, a new model somewhere down the road wants to use that DomainDefinition with different DomainDeployments. We would like a way to facilitate this. One option could be a DomainDefinition(child) that acts exactly like a DomainDefinition, except that it gets its "definition" from some other DomainDefinition. Basically, we need to be able to add any number of DomainDeployments in dependent models that relate back to some original DomainDefinition. I imagine this will raise other questions like "what will we do when the original DomainDefinition changes", but I think this is the right path forward

**Comment by eposse** [comment]
Hi. A couple of questions for clarification:

1. Within a DomainDefinition, every time you create a new DomainDeployment, it will automatically inherit the nodes, etc. from the DomainDefinition, so there is no need to copy and paste a DomainDeployment to the same DomainDefinition. Therefore, I assume that you are attempting to copy and paste DomainDeployments between different DomainDefinitions. Is that right? This may be tricky, since, for example, the nodes in the DomainDeployment being copied may not correspond to the nodes of the DomainDefinition where it is being pasted. If that is what you would like, then what should be the behaviour? For example, if the Deployment to be copied has nodes n1 and n2 and the DomainDefinition where it is going to be copied has nodes n1 and n2, should the receiving DomainDefinition (and all its DomainDeployments) be updated with the nodes n1 and n2? Would the overlapping nodes overwrite existing ones or create new ones (e.g. n1_2)? Overall, copy-and-paste seems to bring up a lot of potential problems (and besides, FYI Papyrus has a history of technical problems with copy and paste).
2.1. By "top level options" do you mean the options when right-clicking the root model element?
2.2. Currently the Generate Descriptors on a Deployment plan will generate the CDP and the CDDs for the old-style Domains in the plan. What you want is to generate the CDDs for those old-style *and* all the DomainDeployments inside any DomainDefinition in the plan as well? If so, should we keep the menu option that generates only the CDP file?
2.3. This is already implemented. The only difference is that the menu label says "Generate All CDD files" but that means all CDD files in that DomainDefinition. Would you like to just simply change the label?

With respect to the new enhancement, I'll have to think a little bit about the best way to implement it, but may I suggest opening that a new separate issue for it?

**Comment by j26151** [comment]
> Within a DomainDefinition, every time you create a new DomainDeployment, it will automatically inherit the nodes, etc. from the DomainDefinition, so there is no need to copy and paste a DomainDeployment to the same DomainDefinition. Therefore, I assume that you are attempting to copy and paste DomainDeployments between different DomainDefinitions. Is that right? This may be tricky, since, for example, the nodes in the DomainDeployment being copied may not correspond to the nodes of the DomainDefinition where it is being pasted. If that is what you would like, then what should be the behaviour? For example, if the Deployment to be copied has nodes n1 and n2 and the DomainDefinition where it is going to be copied has nodes n1 and n3, should the receiving DomainDefinition (and all its DomainDeployments) be updated with the nodes n1 and n2 so that all its Deployments have n1, n2 and n3? Would the overlapping nodes overwrite existing ones or create new ones (e.g. n1_2)? Overall, copy-and-paste seems to bring up a lot of potential problems (and besides, FYI Papyrus has a history of technical problems with copy and paste).

This use case is just for copying DomainDeployments within a DomainDefinition. This could be useful in cases where we have many IP addresses or ports that are the same, but need small tweaks. We do not need to copy DomainDeployments from one DomainDefinition to another.

> 2.1. By "top level options" do you mean the options when right-clicking the root model element?

Root level, yes. Also any other level in the hierarchy that isn't an actual Deployment Plan or DomainDeployment  (just like generating IDL from different levels of the model)

> 2.2. Currently the Generate Descriptors on a Deployment plan will generate the CDP and the CDDs for the old-style Domains in the plan. What you want is to generate the CDDs for those old-style and all the DomainDeployments inside any DomainDefinition in the plan as well? If so, should we keep the menu option that generates only the CDP file?

No, just a single option to generate descriptors. This would generate all deployment artifacts required for that deployment plan. CDPs, CDDs from old-style domains and CDDs from new-style domains.

> 2.3. This is already implemented. The only difference is that the menu label says "Generate All CDD files" but that means all CDD files in that DomainDefinition. Would you like to just simply change the label?

I understand this is already implemented. I just wanted to be clear that this was fine since we are requesting changes for the other two situations.

> With respect to the new enhancement, I'll have to think a little bit about the best way to implement it, but may I suggest opening that a new separate issue for it?

We can create a new issue for this.

Overall, my goal with the above changes is to keep things looking almost the same as they do now, but able to make use of the new functionality of these DomainDefinitions

**Comment by eposse** [comment]
> > Within a DomainDefinition, every time you create a new DomainDeployment, it will automatically inherit the nodes, etc. from 
 
> This use case is just for copying DomainDeployments within a DomainDefinition. This could be useful in cases where we have many IP addresses or ports that are the same, but need small tweaks. We do not need to copy DomainDeployments from one DomainDefinition to another.

Ok I see. I do see that copy-and-paste is not working, but I'm observing a different behaviour. When I copy a DomainDeployment and paste it, at first it appears as if nothing happened, but the copy appears once you edit something else in the model. This is not good. Nevertheless, the copy does have the copy of the contents of the original.

To try to reproduce what you are seeing, how exactly are you copying and pasting? With the keyboard [Ctrl]+[C] -> [Ctrl]+[V], with the context menu? by drag-and-drop?

> > 2.1. By "top level options" do you mean the options when right-clicking the root model element?
> 
> Root level, yes. Also any other level in the hierarchy that isn't an actual Deployment Plan or DomainDeployment (just like generating IDL from different levels of the model)

Ok.
 
> > 2.2. Currently the Generate Descriptors on a Deployment plan will generate the CDP and the CDDs for the old-style Domains in the plan. What you want is to generate the CDDs for those old-style and all the DomainDeployments inside any DomainDefinition in the plan as well? If so, should we keep the menu option that generates only the CDP file?
> 
> No, just a single option to generate descriptors. This would generate all deployment artifacts required for that deployment plan. CDPs, CDDs from old-style domains and CDDs from new-style domains.

Ok.
 
> > 2.3. This is already implemented. The only difference is that the menu label says "Generate All CDD files" but that means all CDD files in that DomainDefinition. Would you like to just simply change the label?
> 
> I understand this is already implemented. I just wanted to be clear that this was fine since we are requesting changes for the other two situations.

Ok

> > With respect to the new enhancement, I'll have to think a little bit about the best way to implement it, but may I suggest opening that a new separate issue for it?
> 
> We can create a new issue for this.

Thanks
 
> Overall, my goal with the above changes is to keep things looking almost the same as they do now, but able to make use of the new functionality of these DomainDefinitions

Right.

**Comment by j26151** [comment]
> Ok I see. I do see that copy-and-paste is not working, but I'm observing a different behaviour. When I copy a DomainDeployment and paste it, at first it appears as if nothing happened, but the copy appears once you edit something else in the model. This is not good. Nevertheless, the copy does have the copy of the contents of the original.

To try to reproduce what you are seeing, how exactly are you copying and pasting? With the keyboard [Ctrl]+[C] -> [Ctrl]+[V], with the context menu? by drag-and-drop?

I did see the same behaviour with needing to edit something else in the model, but I thought maybe it was something else in my environment. When I copy using Ctrl + C -> Ctrl + V, the new domain is missing "parts"
![image](https://user-images.githubusercontent.com/49879646/185431962-b9f826ab-db13-4dab-b186-931972b5007d.png)
and if I try to click on node1 below it, it doesn't have any CX properties for me to edit.

**Comment by j26151** [comment]
Btw, this occurs with using the context menu as well (copy and paste). I have not tried drag and drop.

**Comment by j26151** [comment]
They appear in the model explorer, but if I click on one of them, its empty and I cannot set an ip address or port:
![image](https://user-images.githubusercontent.com/49879646/185434733-8d6fdb35-b2b0-4340-b22d-119b43790a45.png)

**Comment by eposse** [comment]
Ok, I see. I'll look into it. Thanks.

**Comment by eposse** [comment]
There are two immediately observable behaviours which are incorrect when copying and pasting DomainDeployments. The first is that the "parts" of the DomainDeployment stereotype are not copied most of the time (although I observed that one time they were copied, so it may be random). The second is that the DomainDeploymentPart stereotypes of the parts are not copied for the contained parts (nodes, etc.).

As the "parts" association has composite aggregation, this may be related to this Papyrus issue: [Papyrus Bug #552410](https://bugs.eclipse.org/bugs/show_bug.cgi?id=552410). The bug claims to have been solved, but something else is going on. Our code includes a class `com.zeligsoft.domain.dds4ccm.ui.paste.PostStereotypePasteStrategy` which apparently is intended to deal with that Bug, so maybe the problem is there.

I'll continue to investigate.

**Comment by eposse** [comment]
More details: 

There are slightly different problems that occur when copying and pasting, in two scenarios: when copying and pasting the first time, and when pasting a second time without re-copying first. The end result is the same, with parts and stereotypes missing, but Internally the behaviour is actually different.

In all cases the problem arises from our PasteStrategy `com.zeligsoft.domain.dds4ccm.ui.paste.PostStereotypePasteStrategy` and the command `com.zeligsoft.domain.dds4ccm.ui.paste.PostDuplicateStereotypeCommand` it invokes. These where introduced by Young-Soo to deal with a bug in Papyrus: [Bug 561667](https://bugs.eclipse.org/bugs/show_bug.cgi?id=561667). 

However, it looks like this code makes some assumptions that do not always hold. It seems to assume that elements have exactly *one* stereotype applied, but this is not always the case. In our case, the parts inside a `DomainDeployment` have two stereotypes (`DeploymentPart` and `DomainDeploymentPart`, the later introduced because it was needed for code generation). When the part is first pasted, a key part of the code (`com.zeligsoft.domain.dds4ccm.ui.paste.PostDuplicateStereotypeCommand.reloadInTargetContext(ResourceSet, EObject)`) returns the first such stereotype (`DeplymentPart`) when it should be returning the second (`DomainDeploymentPart`) since the reference being copied (`DomainDeployment.parts`) is a list of elements with the second stereotype. 

Then, if the object is pasted for a second time without being re-copied, the same code above fails, as the parts no longer have a stereotype applied to them, so attempting to get the first stereotype in the list raises an exception.

**Comment by ysroh** [comment]
The assumption was valid for CX until the new stereotype was introduced. You are welcome to modify the code to fix the issue.

**Comment by eposse** [comment]
@j26151 Ok, I've addressed all issues and it is working now (including copy-and-paste) and the menus as requested. You can get the rpm [here](https://github.com/ZeligsoftDev/CX4CBDDS/suites/7990155442/artifacts/342122206). Let me know if it addresses everything.

**Comment by ysroh** [comment]
I wonder what causes the problem with copy/paste, but I would be cautious that the new fix does not cause any problems during the cross-model copy/paste. You should be able to copy the object from one model and paste it onto another model.

**Comment by eposse** [comment]
The problem with copy and paste was that the code had a long-standing assumption that elements would have only one stereotype and I had introduced an additional stereotype to deal with an issue in the code generator. The new fixes remove the new stereotype and update the code generator so that it can work without it. Copy and paste works now in the latest fix.

**Comment by eposse** [comment]
I'll investigate.

**Comment by j26151** [comment]
I was able to recreate it. I found a way to get a larger stack trace:
![image](https://user-images.githubusercontent.com/49879646/187251750-b18eebbe-b5c4-47a9-b5f5-6c1c4fda3be2.png)
It appears to happen anytime there are delegated ports on assemblies.

**Comment by eposse** [comment]
Great! And just to confirm, this exception does occurs only with the version from this PR, but not with the main papyrus branch, is that right?

**Comment by j26151** [comment]
Good point. I just downgraded my version to one that is the latest from the main papyrus branch. The issue is still happening. It appears that this issue does not have anything to do with your branch, rather its a bug that has been hanging around for who knows how long. I know for a fact the maintenance stream does not have this issue. I do not know when this issue was introduced, but it is a show stopper for us.

**Comment by eposse** [comment]
Ok, so perhaps we could have a separate issue for that? Hopefully with a small reproducer model, as I am unsure how to create delegated ports.

It's possible that the duplicated properties in the deployment plan happens only with this branch, so I'll investigate that as well.

**Comment by j26151** [comment]
To delegate a port, all you do is create an assembly with a port on it that is the same port as a component inside of the assembly. Connect the two, and you get an assmebly that delegates its ports to an internal component:
![image](https://user-images.githubusercontent.com/49879646/187256791-76a91ead-9018-48bf-99b6-9219cb37fc25.png)
and 
![image](https://user-images.githubusercontent.com/49879646/187256893-b3f4e80a-f57b-4e2a-b37d-20471681ccac.png)

As for creating a separate issue, we can definitely create a new issue to track this, but I will not be able to verify or approve this issue without that fix (either fixed on this issue, or this issue would need to be rebased on whatever issue fixes it)

**Comment by j26151** [comment]
@eposse  I see you pushed some commits up a few days ago, is there anything you would like me to test for this?

**Comment by eposse** [comment]
Hi @j26151. Those commits don't yet fix the code generation issue. They addressed some concerns Young-Soo had about copy and paste. 

I have found the cause of the code generation issue, and it was a regression that happened when we dealt with issue #397. Delegated ports was a case that we didn't consider when dealing with that issue. 

I also have a solution, but I'm not yet certain it covers all cases, which is what I'm investigating today. That issue is in fact, as I said before, unrelated to this issue, and the affected code is separate from the code in this issue, so I am tempted to put it in a separate Issue and PR. 

In short, you can test the new one, but you won't see much of a difference, but the solution to the codegen issue will be ready soon. Thanks for your patience.

**Comment by eposse** [comment]
I've opened Issue #428 to deal with the code generation regression. See the comments there.

**Comment by j26151** [comment]
@eposse Have you been able to get around merging/rebasing this branch? If so I can build and test it to close this out (or at least provide more feedback)

**Comment by eposse** [comment]
@j26151 There was indeed a conflict with the code committed for dealing with Issue #423, but I've been able to merge it. Testing it looks good, but there is still a problem. When selecting an element in the Deployment editor, the properties view is incorrectly rendered, and shows the "Properties" field twice, sometimes one below the other, sometimes at its side, as shown in the screenshot below. I'm going to investigate. Let me know if you'd like me to push the latest commit after the merge, or if you'd rather wait until the properties view issue is resolved.

![Screenshot from 2022-09-12 15-00-29](https://user-images.githubusercontent.com/3011155/189735335-803e567e-a451-4072-8ae6-0015299ba77f.png)

**Comment by ysroh** [comment]
I get the following exception and this will cause the properties view to be corrupted and it will not render it correctly after the exception.
!STACK 0
java.lang.NullPointerException
	at org.eclipse.uml2.common.util.CacheAdapter.getCacheAdapter(CacheAdapter.java:249)
	at com.zeligsoft.base.zdl.util.ZDLUtil$ZDLConceptInfo.ensureCacheAdapter(ZDLUtil.java:3100)
	at com.zeligsoft.base.zdl.util.ZDLUtil$ZDLConceptInfo.get(ZDLUtil.java:3079)
	at com.zeligsoft.base.zdl.util.ZDLUtil.getZDLConcepts(ZDLUtil.java:212)
	at com.zeligsoft.base.zdl.util.ZDLUtil.isZDLConcept(ZDLUtil.java:1442)
	at com.zeligsoft.domain.idl3plus.utils.IDL3PlusUtil.getDataSpaceFromPerPort(IDL3PlusUtil.java:1001)
	at com.zeligsoft.domain.idl3plus.ui.providers.IDL3PlusPropertyContentProvider.getChildren(IDL3PlusPropertyContentProvider.java:94)
	at org.eclipse.jface.viewers.AbstractTreeViewer.getRawChildren(AbstractTreeViewer.java:1429)

**Comment by eposse** [comment]
I've resolved the NPE, and while the properties view is not "corrupted" anymore, the properties are still duplicated. See the screenshot. I'll keep investigating.

![Screenshot from 2022-09-13 18-05-14](https://user-images.githubusercontent.com/3011155/190017850-7f506448-74ea-4450-943d-1c9c94356219.png)

**Comment by eposse** [comment]
@j26151 I've resolved the duplicate properties view issue and it's ready for you to test. I think it now addresses all issues.

You can get the RPM from [here](https://github.com/ZeligsoftDev/CX4CBDDS/actions/runs/3063756448).

**Comment by j26151** [comment]
@eposse I tested a good amount and for the most part everything looks good. I did notice a new bug that I believe is related to this branch of work (i.e. it did not appear on the HEAD of papyrus: https://github.com/ZeligsoftDev/CX4CBDDS/commit/248e98795af690ea16248c5bac3b8220cba968ed)

When I open any assembly diagram that has an assembly inside of it, I get errors from eclipse trying to navigate into other assembly diagrams. You can reproduce this by opening our BasicPubSub model and opening the "context" diagram (it has context in the diagram name) then double clicking on the only assembly inside of the diagram. You will see a new tab created in the bottom but the view won't switch and errors are thrown in the error log.

**Comment by eposse** [comment]
Hmm. I am not able to reproduce it. First, I'm not sure we have the same BasicPubSub, as the one I have, from the git repo, has no "context" diagram. See the attached screenshot.

Nevertheless, I tried it with another model that has nested assemblies, (see the attached model), and I am able to double-click on parts whose type is an assembly and navigate to them without trouble, and no exceptions.

[axPubSub.zip](https://github.com/ZeligsoftDev/CX4CBDDS/files/9600936/axPubSub.zip)


![Screenshot from 2022-09-19 13-00-55](https://user-images.githubusercontent.com/3011155/191072890-1572f74b-f8a1-4d83-863f-8abb2ecb9fbf.png)

**Comment by j26151** [comment]
Its likely that you have an older version where we didn't have composite structure diagrams created yet. If you create a new composite structure diagram for BasicPubSub_asm_context, it should contain inside of it a single assembly for BasicPubSub_asm. If, in the diagram itself you double clicked on the BasicPubSub_asm, this is where I would see the issue.

**Comment by eposse** [comment]
Still can't reproduce it. Here's what I did: 

1) started CX on a new workspace
2) imported (copying) BasicPubSub
3) opened and migrated the model and saved
4) navigated to assemblies/BasicPubSub_asm_context; right-click -> New Diagram -> Composite Structure Diagram; saved
5) the diagram is open; double-clicked on the only part in the context assembly.

Result: it opened the BasicPubSub_asm diagram as expected; no exceptions are raised; no errors are logged.

I observe the same behaviour with the axPubSub example I sent in the previous message. 

Are you certain you are testing with the latest build from this branch on a fresh workspace? 

Can you test also with the model I sent?

**Comment by j26151** [comment]
I was certainly on the latest build from this branch, but not a fresh workspace. In any case, I can no longer reproduce the problem either. Papyrus must have just gotten in a bad state and needed a restart. I think for the time being we can approve and merge this issue in and if we see this issue crop up again we will have to try to come up with a more reliable reproducer for it.

**Comment by eposse** [comment]
Great! Thanks. I'll merge it then.

**Comment by eposse** [comment]
Closing it.
